File: ThanosSort/thanos_sort.py
import random


def is_sorted(array):
    for i in range(1, len(array)):
        if array[i - 1] > array[i]:
            return False
    return True


def thanos_sort(universe):
    if is_sorted(universe):
        return universe

    size = len(universe)
    indexes = range(0, size)
    half = int(size / 2)

    to_die = random.sample(indexes, half)
    universe = [i for j, i in enumerate(universe) if j not in to_die]

    return thanos_sort(universe)


def main():
    random_universe = random.sample(range(1, 1000), 100)
    print("Universe: ", random_universe)

    ordered_universe = thanos_sort(random_universe)
    print("Ordered universe: ", ordered_universe)


if __name__ == "__main__":
    main()

File: ThanosSort/thanos_sort.cpp
#include <iostream>
#include <vector>
#include <ctime>
#include <cstdlib>

using namespace std;

// Time complexity: O(n)
vector<int> randomUniverse() {
  vector<int> universe;
  srand(time(NULL));
  for(int i = 0; i < 100; i++) {
    universe.push_back(rand() % 1000);
  }
  return universe;
}

// Time complexity: O(n)
void printUniverse(string universeName, vector<int> universe) {
  int universeSize = universe.size();
  cout << universeName << ": [" << universe[0];
  for(int i = 1; i < universeSize; i++) {
    cout << ", " << universe[i];
  }
  cout << "]" << endl;
}

// Time complexity: O(n)
bool isSorted(vector<int> array) {
  int size = array.size();
  for(int i = 1; i < size; i++) {
    if(array[i - 1] > array[i]) {
      return false;
    }
  }
  return true;
}

// Time complexity: O(n)
void killRandomElement(vector<int>& universe) {
  int index = rand() % universe.size();
  universe.erase(universe.begin() + index);
}

// Time complexity: O(n ^ 2)
void killHalfUniverse(vector<int>& universe) {
  int halfUniverseSize = universe.size() / 2;
  for(int i = 0; i < halfUniverseSize; i++) {
    killRandomElement(universe);
  }
}

// Time complexity: O(n ^ 2 * log n)
void thanosSort(vector<int>& universe) {
  while(!isSorted(universe)) {
    killHalfUniverse(universe);
  }
}

// Time complexity: O(n ^ 2 * log n)
int main() {
  vector<int> universe = randomUniverse();
  printUniverse("Universe", universe);
  thanosSort(universe);
  printUniverse("Ordered universe: ", universe);
  return 0;
}
